### PR TITLE
XmlLayout - Added object reflection using ObjectReflectionCache

### PR DESCRIPTION
**Pull request by snakefoot**
Continue the work from #2670 and #2989

### CODE REVIEW

File: src/NLog/Layouts/XmlLayout.cs [hunk 1]
```diff
@@ -381,38 +395,188 @@ private void AppendLogEventXmlProperties(LogEventInfo logEventInfo, StringBuilde
...
-                    AppendXmlPropertyValue(key, propertyValue, sb, sb.Length == orgLength);
+                    AppendXmlPropertyValue(key, propertyValue, Convert.GetTypeCode(propertyValue), sb, orgLength);
                 }
             }
 #endif
 
             if (IncludeAllProperties && logEventInfo.HasProperties)
             {
-                foreach (var property in logEventInfo.Properties)
+                var propertiesList = logEventInfo.CreateOrUpdatePropertiesInternal(true) as IEnumerable<MessageTemplates.MessageTemplateParameter>;
```
Comment (304NotModified) [inline src/NLog/Layouts/XmlLayout.cs]: why the `as`?

Cant we use  `IEnumerable<MessageTemplates.MessageTemplateParameter> propertiesList = logEventInfo.CreateOrUpdatePropertiesInternal(true)` (runtime safe)

Comment (snakefoot) [reply src/NLog/Layouts/XmlLayout.cs]: PropertiesDictionary simulates two IEnumerables. The old LogEventInfo.Properties (`IDictionary<object, object>`) and the new LogEventInfo.MessageTemplateParameters (`IEnumerable<MessageTemplateParameter>`). I want to enumerate the new one to access CaptureType.

Comment (304NotModified) [reply src/NLog/Layouts/XmlLayout.cs]: Yes I understand, 

but we can use `IEnumerable<MessageTemplates.MessageTemplateParameter> propertiesList = logEventInfo.CreateOrUpdatePropertiesInternal(true)` isn't?

Comment (snakefoot) [reply src/NLog/Layouts/XmlLayout.cs]: Yes you are right. Fixed

**Comment by snakefoot** [comment]
Btw. regarding the "performance"-label then it is more an enhancement. Will probably introduce a minor performance hit :)

File: src/NLog/Layouts/XmlLayout.cs (hunk 1 above)
Comment (304NotModified) [reply src/NLog/Layouts/XmlLayout.cs]: 🎉

**Comment by 304NotModified** [comment]
Thanks! Merged!

**Pull request merged**